spec/models/spree/adyen/notification_processor_spec.rb

```diff
@@ -148,21 +148,40 @@
     context "when event is REFUND" do
       include_examples "processed event"
 
-      let(:event_type) { :refund }
-      let(:payment_state) { "processing" }
+      context "when refunded from Solidus"
+        let(:event_type) { :refund }
+        let(:payment_state) { "processing" }
+
+        it "creates a refund" do
+          expect { subject }.
+            to change { payment.reload.refunds.count }.
+            from(0).
+            to(1)
+        end
 
-      it "creates a refund" do
-        expect { subject }.
-          to change { payment.reload.refunds.count }.
-          from(0).
-          to(1)
+        it "changes the payment state to completed" do
+          expect { subject }.
+            to change { payment.reload.state }.
+            from("processing").
+            to("completed")
+        end
       end
 
-      it "changes the payment state to completed" do
-        expect { subject }.
-          to change { payment.reload.state }.
-          from("processing").
-          to("completed")
+      context "when refunded from Adyen" do
+        let(:event_type) { :refund }
+        let(:payment_state) { "completed" }
+
+        it "creates a refund" do
+          expect { subject }.
+            to change { payment.reload.refunds.count }.
+            from(0).
+            to(1)
+        end
+
+        it "changes the payment state to completed" do
+          expect { subject }.
+            not_to change { payment.reload.state }.
+            from("completed")
```

Comment:
These two examples, and could be moved into an shared example, then this and the previous context become
```ruby
shared_examples "refund" do
  let(:event_type) { :refund }
  it "creates a refund" do
    expect { subject }.
      to change { payment.reload.refunds.count }.
      from(0).
      to(1)
  end
end

context "when refunded from Solidus" do
  let(:payment_state) { "processing" }
  include_examples "refund"

  it "changes the payment state to completed" do
    expect { subject }.
      to change { payment.reload.state }.
      from("processing").
      to("completed")
  end
end

context "when refunded from Adyen" do
  let(:payment_state) { "completed" }
  include_examples "refund"

  it "does not change the payment state" do
    expect { subject }.
      not_to change { payment.reload.state }.
      from("completed")
  end
end
```